File: lib/models/player_configuration.dart
import 'dart:core';

import 'package:udevs_video_player/models/programs_info.dart';
import 'package:udevs_video_player/models/season.dart';

import 'tv_categories.dart';

export 'programs_info.dart';
export 'season.dart';
export 'tv_categories.dart';
export 'tv_channel.dart';

class PlayerConfiguration {
  const PlayerConfiguration({
    required this.initialResolution,
    required this.resolutions,
    required this.qualityText,
    required this.speedText,
    required this.lastPosition,
    required this.title,
    required this.isSerial,
    required this.episodeButtonText,
    required this.nextButtonText,
    required this.showController,
    required this.playVideoFromAsset,
    required this.assetPath,
    required this.seasonIndex,
    required this.episodeIndex,
    required this.isMegogo,
    required this.isMoreTv,
    required this.isPremier,
    required this.videoId,
    required this.sessionId,
    required this.megogoAccessToken,
    required this.authorization,
    required this.autoText,
    required this.baseUrl,
    required this.fromCache,
    required this.movieShareLink,
    required this.userId,
    required this.profileId,
    required this.noneText,
    required this.subtitleText,
    required this.audioText,
    required this.defaultText,
    this.seasons = const [],
    this.isLive = false,
    this.tvProgramsText = 'Tv Programs',
    this.programsInfoList = const [],
    this.ip = '',
    this.selectChannelIndex = 0,
    this.selectTvCategoryIndex = 0,
    this.tvCategories = const [],
    this.sendMovieTrack = true,
  });

  final Map<String, String> initialResolution;
  final Map<String, String> resolutions;
  final String qualityText;
  final String speedText;
  final int lastPosition;
  final String title;
  final bool isSerial;
  final String episodeButtonText;
  final String nextButtonText;
  final List<Season> seasons;
  final bool isLive;
  final String tvProgramsText;
  final List<ProgramsInfo> programsInfoList;
  final bool showController;
  final bool playVideoFromAsset;
  final String assetPath;
  final int seasonIndex;
  final int episodeIndex;
  final bool isMegogo;
  final bool isPremier;
  final bool isMoreTv;
  final String videoId;
  final String sessionId;
  final String megogoAccessToken;
  final String authorization;
  final String autoText;
  final String baseUrl;
  final String movieShareLink;
  final bool fromCache;
  final List<TvCategories> tvCategories;
  final String ip;
  final int selectChannelIndex;
  final int selectTvCategoryIndex;
  final String userId;
  final String profileId;
  final String noneText;
  final String subtitleText;
  final String audioText;
  final String defaultText;
  final bool sendMovieTrack;

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{};
    map['initialResolution'] = initialResolution;
    map['resolutions'] = resolutions;
    map['qualityText'] = qualityText;
    map['speedText'] = speedText;
    map['lastPosition'] = lastPosition;
    map['title'] = title;
    map['isSerial'] = isSerial;
    map['episodeButtonText'] = episodeButtonText;
    map['nextButtonText'] = nextButtonText;
    map['seasons'] = seasons.map((v) => v.toJson()).toList();
    map['isLive'] = isLive;
    map['tvProgramsText'] = tvProgramsText;
    map['programsInfoList'] = programsInfoList.map((v) => v.toJson()).toList();
    map['showController'] = showController;
    map['playVideoFromAsset'] = playVideoFromAsset;
    map['assetPath'] = assetPath;
    map['seasonIndex'] = seasonIndex;
    map['episodeIndex'] = episodeIndex;
    map['isMegogo'] = isMegogo;
    map['isMoreTv'] = isMoreTv;
    map['isPremier'] = isPremier;
    map['videoId'] = videoId;
    map['sessionId'] = sessionId;
    map['megogoAccessToken'] = megogoAccessToken;
    map['authorization'] = authorization;
    map['autoText'] = autoText;
    map['baseUrl'] = baseUrl;
    map['fromCache'] = fromCache;
    map['movieShareLink'] = movieShareLink;
    map['ip'] = ip;
    map['selectChannelIndex'] = selectChannelIndex;
    map['selectTvCategoryIndex'] = selectTvCategoryIndex;
    map['profileId'] = profileId;
    map['userId'] = userId;
    map['tvCategories'] = tvCategories.map((v) => v.toJson()).toList();
    map['noneText'] = noneText;
    map['subtitleText'] = subtitleText;
    map['audioText'] = audioText;
    map['defaultText'] = defaultText;
    map['sendMovieTrack'] = sendMovieTrack;
    return map;
  }

  @override
  String toString() => 'PlayerConfiguration{'
      'initialResolution: $initialResolution, '
      'resolutions: $resolutions, '
      'qualityText: $qualityText, '
      'speedText: $speedText, '
      'lastPosition: $lastPosition, '
      'title: $title, '
      'isSerial: $isSerial, '
      'episodeButtonText: $episodeButtonText, '
      'nextButtonText: $nextButtonText, '
      'seasons: $seasons, '
      'isLive: $isLive, '
      'tvProgramsText: $tvProgramsText, '
      'programsInfoList: $programsInfoList, '
      'showController: $showController, '
      'playVideoFromAsset: $playVideoFromAsset, '
      'assetPath: $assetPath, '
      'seasonIndex: $seasonIndex, '
      'episodeIndex: $episodeIndex, '
      'isMegogo: $isMegogo, '
      'isMoreTv: $isMoreTv, '
      'isPremier: $isPremier, '
      'videoId: $videoId, '
      'sessionId: $sessionId, '
      'megogoAccessToken: $megogoAccessToken, '
      'authorization: $authorization, '
      'autoText: $autoText '
      'baseUrl: $baseUrl, '
      'fromCache: $fromCache, '
      'movieShareLink: $movieShareLink, '
      'channels: $tvCategories, '
      'ip: $ip'
      'selectChannelIndex: $selectChannelIndex'
      'userId: $userId'
      'profileId: $profileId'
      'noneText: $noneText'
      'subtitleText: $subtitleText'
      'audioText: $audioText'
      'defaultText: $defaultText'
      'sendMovieTrack: $sendMovieTrack'
      '}';
}

File: ios/Classes/Player/Models/PlayerConfiguration.swift
//
//  PlayerConfiguration.swift
//  udevs_video_player
//
//  Created by Udevs on 08/10/22.
//

import Foundation

struct PlayerConfiguration{
    var initialResolution: [String:String]
    var resolutions: [String:String]
    var url, qualityText: String
    var speedText: String
    var lastPosition: Int
    var title: String
    var isSerial: Bool
    var episodeButtonText: String
    var nextButtonText: String
    var seasons: [Season]
    var isLive: Bool
    var tvProgramsText: String
    var programsInfoList: [ProgramInfo]
    var showController: Bool
    var playVideoFromAsset: Bool
    var assetPath: String?
    var seasonIndex: Int
    var episodeIndex: Int
    var isMegogo: Bool
    var isPremier: Bool
    var isMoreTv: Bool
    var videoId: String
    var sessionId: String
    var megogoAccessToken: String
    var authorization: String
    var autoText: String
    var baseUrl: String
    var movieShareLink: String
    var ip: String
    var selectChannelIndex: Int
    var selectTvCategoryIndex: Int
    var tvCategories: [TvCategories]
    var noneText: String
    var defaultText: String
    var profileID:String
    var userID:String
    var sendMovieTrack:Bool
    var shareText:String
    var subtitleText:String
    var audioText:String
    
    init(initialResolution: [String : String], resolutions: [String : String], qualityText: String, speedText: String, lastPosition: Int, title: String, isSerial: Bool, episodeButtonText: String, nextButtonText: String, seasons: [Season], isLive: Bool, tvProgramsText: String, programsInfoList: [ProgramInfo], showController: Bool, playVideoFromAsset: Bool, assetPath: String? = nil, seasonIndex: Int, episodeIndex: Int, isMegogo: Bool, isPremier: Bool, isMoreTv: Bool, videoId: String, sessionId: String, megogoAccessToken: String, authorization: String, autoText: String, baseUrl: String,url: String,movieShareLink: String, ip : String, selectChannelIndex: Int, selectTvCategoryIndex: Int, tvCategories: [TvCategories],noneText:String,defaultText:String,profileID:String,userID:String,sendMovieTrack:Bool,shareText:String,subtitleText:String,
     audioText:String) {
        self.initialResolution = initialResolution
        self.resolutions = resolutions
        self.qualityText = qualityText
        self.speedText = speedText
        self.lastPosition = lastPosition
        self.title = title
        self.isSerial = isSerial
        self.episodeButtonText = episodeButtonText
        self.nextButtonText = nextButtonText
        self.seasons = seasons
        self.isLive = isLive
        self.tvProgramsText = tvProgramsText
        self.programsInfoList = programsInfoList
        self.showController = showController
        self.playVideoFromAsset = playVideoFromAsset
        self.assetPath = assetPath
        self.seasonIndex = seasonIndex
        self.episodeIndex = episodeIndex
        self.isMegogo = isMegogo
        self.isPremier = isPremier
        self.isMoreTv = isMoreTv
        self.videoId = videoId
        self.sessionId = sessionId
        self.megogoAccessToken = megogoAccessToken
        self.authorization = authorization
        self.autoText = autoText
        self.baseUrl = baseUrl
        self.url = url
        self.movieShareLink = movieShareLink
        self.ip = ip
        self.selectChannelIndex = selectChannelIndex
        self.selectTvCategoryIndex = selectTvCategoryIndex
        self.tvCategories = tvCategories
        self.noneText = noneText
        self.defaultText = defaultText
        self.profileID = profileID
        self.userID = userID
        self.sendMovieTrack = sendMovieTrack
        self.shareText = shareText
        self.subtitleText = subtitleText
        self.audioText = audioText
    }
    
    static func fromMap(map : [String:Any])->PlayerConfiguration {
        var season : [Season] = []
        var programInfos: [ProgramInfo] = []
        var tvCategories: [TvCategories] = []
        var programsInfoListMap : [Dictionary<String, Any>]?
        var tvCategoriesMap : [Dictionary<String, Any>]?
        var seasonsMap : [Dictionary<String, Any>]?
        programsInfoListMap = map["programsInfoList"] as? [Dictionary<String, Any>]
        seasonsMap = map["seasons"] as? [Dictionary<String, Any>]
        tvCategoriesMap = map["tvCategories"] as? [Dictionary<String, Any>]
        programsInfoListMap?.forEach({ data in
            let program = ProgramInfo.fromMap(map: data)
            programInfos.append(program)
        })
        seasonsMap?.forEach({ data in
            let program = Season.fromMap(map: data)
            season.append(program)
        })
        tvCategoriesMap?.forEach({ data in
            let program = TvCategories.fromMap(map: data)
            tvCategories.append(program)
        })
        
        return PlayerConfiguration(initialResolution: map["initialResolution"] as! [String:String],
                                   resolutions: map["resolutions"] as! [String:String],
                                   qualityText: map["qualityText"] as! String,
                                   speedText: map["speedText"] as! String,
                                   lastPosition: map["lastPosition"] as! Int,
                                   title: map["title"] as! String,
                                   isSerial: map["isSerial"] as! Bool,
                                   episodeButtonText: map["episodeButtonText"] as! String,
                                   nextButtonText: map["nextButtonText"] as! String,
                                   seasons: season,
                                   isLive: map["isLive"] as! Bool,
                                   tvProgramsText: map["tvProgramsText"] as! String,
                                   programsInfoList: programInfos,
                                   showController : map["showController"] as! Bool,
                                   playVideoFromAsset : map["playVideoFromAsset"] as! Bool,
                                   assetPath:map["assetPath"] as? String,
                                   seasonIndex: map["seasonIndex"] as! Int,
                                   episodeIndex: map["episodeIndex"] as! Int,
                                   isMegogo: map["isMegogo"] as! Bool,
                                   isPremier: map["isPremier"] as! Bool,
                                   isMoreTv: map["isMoreTv"] as! Bool,
                                   videoId: map["videoId"] as! String,
                                   sessionId: map["sessionId"] as! String,
                                   megogoAccessToken: map["megogoAccessToken"] as! String,
                                   authorization: map["authorization"] as! String,
                                   autoText: map["autoText"] as! String,
                                   baseUrl: map["baseUrl"] as! String,
                                   url: (map["initialResolution"] as! [String:String]).values.first ?? "",
                                   movieShareLink: map["movieShareLink"] as! String,
                                   ip: map["ip"] as! String,
                                   selectChannelIndex: map["selectChannelIndex"] as? Int ?? 0, selectTvCategoryIndex: map["selectTvCategoryIndex"] as? Int ?? 0,
                                   tvCategories: tvCategories,
                                   noneText: map["noneText"] as! String,
                                   defaultText: map["defaultText"] as! String,
                                   profileID: map["profileId"] as! String,
                                   userID: map["userId"] as! String,
                                   sendMovieTrack: map["sendMovieTrack"] as! Bool,
                                   shareText: map["shareText"] as! String,
                                   subtitleText: map["subtitleText"] as! String,
                                   audioText: map["audioText"] as! String
        )
    }
}


struct Season {
    var title: String?
    var movies: [Movie]
    init(title: String? = nil, movies: [Movie]) {
        self.title = title
        self.movies = movies
    }
    
    static func fromMap(map : [String:Any])->Season{
        var s: [Movie] = []
        var movies: [Dictionary<String, Any>]?
        movies = map["movies"] as! [Dictionary<String, Any>]?
        movies?.forEach { data in
            let movi = Movie.fromMap(map: data)
            s.append(movi)
        }
        return Season(title:  map["title"] as? String, movies: s)
    }
}

struct Movie {
    var id: String?
    var title: String?
    var description: String?
    var image: String?
    var duration: Int?
    var resolutions: [String:String]
    
    init(id: String? = nil, title: String? = nil, description: String? = nil, image: String? = nil, duration: Int? = nil, resolutions: [String : String]) {
        self.id = id
        self.title = title
        self.description = description
        self.image = image
        self.duration = duration
        self.resolutions = resolutions
    }
    
    static func fromMap(map : [String:Any])->Movie{
        return Movie(id: (map["id"] as? String),title: (map["title"] as? String), description: map["description"] as? String, image: (map["image"] as? String), duration: (map["duration"] as? Int), resolutions: (map["resolutions"] as! [String:String]))
    }
}
//
//struct ProgramInfo {
//    var day: String
//    var tvPrograms: [TvProgram]?
//    init(day: String, tvPrograms: [TvProgram]? = nil) {
//        self.day = day
//        self.tvPrograms = tvPrograms
//    }
//    static func fromMap(map : [String:Any])->ProgramInfo{
//        var tv: [TvProgram] = []
//        var tvPrograms: [Dictionary<String, Any>]?
//        tvPrograms = map["tvPrograms"] as! [Dictionary<String, Any>]?
//        tvPrograms?.forEach { data in
//            let tvProgram = TvProgram.fromMap(map: data as! [String:String])
//            tv.append(tvProgram)
//        }
//        return ProgramInfo(day: map["day"] as! String,tvPrograms: tv )
//    }
//}
//
//struct TvProgram{
//    var scheduledTime: String?
//    var programTitle: String?
//    init(scheduledTime: String? = nil, programTitle: String? = nil) {
//        self.scheduledTime = scheduledTime
//        self.programTitle = programTitle
//    }
//    
//    static func fromMap(map : [String:String])->TvProgram{
//        return TvProgram(scheduledTime: map["scheduledTime"]!, programTitle: map["programTitle"]!)
//    }
//}
//
//struct Channel {
//    var id: String?
//    var image: String?
//    var name: String?
//    var resolutions: [String:String]
//    
//    init(id: String? = nil, name: String? = nil, image: String? = nil, resolutions: [String : String]) {
//        self.id = id
//        self.name = name
//        self.image = image
//        self.resolutions = resolutions
//    }
//    
//    static func fromMap(map : [String:Any])-> Channel{
//        return Channel(id: (map["id"] as? String),name:(map["name"] as? String), image: (map["image"] as? String), resolutions: (map["resolutions"] as! [String:String]))
//    }
//}
//
//extension Channel: Equatable {
//     static func == (lhs: Channel, rhs: Channel) -> Bool {
//         lhs.id == rhs.id && lhs.name == rhs.name && lhs.image == rhs.image
//     }
// }
//struct TvCategories{
//    var id: String?
//    var title: String?
//    var channels: [Channel]
//    
//    init(id: String? = nil, title: String? = nil, channels: [Channel]) {
//        self.id = id
//        self.title = title
//        self.channels = channels
//    }
//    
//    static func fromMap(map : [String:Any])-> TvCategories{
//        var channels : [Channel] = []
//        var channelsMap : [Dictionary<String, Any>]?
//        channelsMap = map["tvChannels"] as? [Dictionary<String, Any>]
//        channelsMap?.forEach({ data in
//            let program = Channel.fromMap(map: data)
//            channels.append(program)
//        })
//        return TvCategories(id: (map["id"] as? String),title: map["title"] as? String, channels: channels)
//    }
//}

extension PlayerConfiguration{
    func isUzdMovie()->Bool{
        return !(self.isMegogo || self.isPremier || self.isMoreTv)
    }
}
